File: src/Client.UWP/View/Window.cs
using Client.Domain;
using System.Collections.Generic;
using System;

namespace Client.View
{
    /// <summary>
    /// Contains all entities which need to be visible in Game window and converts them
    /// to textures.
    /// </summary>
    public sealed class Window
    {
        private readonly Dictionary<GeoPoint, LocationType> points = new Dictionary<GeoPoint, LocationType>();

        // defines single texture generator strategy
        // area: small array 3*3, where central point defines the point which need th have
        // generated texture, and the rest of array items defines neighbers with indexes as constatns above.
        // if strategy can't generate texture, returns defaultTexture.
        private delegate TextureHolder Generator(LocationType[] area, TextureHolder defaultTexture);

        private WaterTextures water;
        private TextureHolder ground;
        private TextureHolder city;
        private List<Generator> functionStrategies = new List<Generator>();
        private ITileStrategy[] strategies = new ITileStrategy[0];
        private ITileStrategy fallbackStrategy;

        public Window(WaterTextures water, TextureHolder ground, TextureHolder city, ITileStrategy fallbackStrategy, params ITileStrategy[] strategies)
        {
            this.water = water;
            this.ground = ground;
            this.city = city;

            functionStrategies.Add(StrategyForGround);
            functionStrategies.Add(CoastWithLandToTheNorth);
            functionStrategies.Add(CoastWithLandToTheSouth);
            functionStrategies.Add(CoastWithLandToTheWest);
            functionStrategies.Add(CoastWithLandToTheEast);
            functionStrategies.Add(CoastWithLandToTheNorthEast);
            functionStrategies.Add(CoastWithLandToTheNorthWest);
            functionStrategies.Add(CoastWithLandToTheSouthEast);
            functionStrategies.Add(CoastWithLandToTheSouthWest);

            this.strategies = strategies;
            this.fallbackStrategy = fallbackStrategy;
        }

        public void AddIsland(IslandEntity island)
        {
            foreach (var item in island.GeneratePoints())
            {
                var point = new GeoPoint(item.X, item.Y);
                points.Add(point, LocationType.Ground);
            }
        }

        /// <summary>
        /// Defines a city on the map.
        /// 
        /// Needs to be invoked later then <see cref="AddIsland(IslandEntity)"/>
        /// </summary>
        /// <param name="entity"></param>
        public void AddCity(CityEntity entity)
        {
            var point = new GeoPoint(entity.X, entity.Y);
            points[point] = LocationType.City;
        }

        public void Include(LandUnitEntity entity)
        {
            var point = new GeoPoint(entity.X, entity.Y);
            points[point] = LocationType.LandUnit;
        }

        /// <summary>
        /// Generates list of textures for given square, when left bottom corner is provided.
        /// </summary>
        /// <param name="lbx"></param>
        /// <param name="lby"></param>
        /// <param name="dx"></param>
        /// <param name="dy"></param>
        /// <returns>Sequence contains textures, starting from left bottom corner and returning rows first.</returns>
        public IEnumerable<PointContext> GetWindow(int lbx, int lby, int dx, int dy)
        {
            for (int y = 0; y < dy; y++)
                for (int x = 0; x < dx; x++)
                {
                    var geox = lbx + x;
                    var geoy = lby + y;

                    var area = new LocationType[9];
                    var centerOfArea = new GeoPoint(geox, geoy);

                    for (int i = 0; i < 9; i++) area[i] = LocationType.Water;

                    area[Directions.NeighborTopLeft] = points.ContainsKey(centerOfArea.TopLeft()) ? points[centerOfArea.TopLeft()] : LocationType.Water;
                    area[Directions.NeighborNorth] = points.ContainsKey(centerOfArea.Top()) ? points[centerOfArea.Top()] : LocationType.Water;
                    area[Directions.NeighborTopRight] = points.ContainsKey(centerOfArea.TopRight()) ? points[centerOfArea.TopRight()] : LocationType.Water;
                    area[Directions.NeighborWest] = points.ContainsKey(centerOfArea.Left()) ? points[centerOfArea.Left()] : LocationType.Water;
                    area[Directions.NeighborThis] = points.ContainsKey(centerOfArea) ? points[centerOfArea] : LocationType.Water;
                    area[Directions.NeighborEast] = points.ContainsKey(centerOfArea.Right()) ? points[centerOfArea.Right()] : LocationType.Water;
                    area[Directions.NeighborDownLeft] = points.ContainsKey(centerOfArea.DownLeft()) ? points[centerOfArea.DownLeft()] : LocationType.Water;
                    area[Directions.NeighborSouth] = points.ContainsKey(centerOfArea.Down()) ? points[centerOfArea.Down()] : LocationType.Water;
                    area[Directions.NeighborDownRight] = points.ContainsKey(centerOfArea.DownRight()) ? points[centerOfArea.DownRight()] : LocationType.Water;

                    var centerTexture = new TextureHolder();
                    var handled = false;
                    foreach (var strategy in functionStrategies)
                    {
                        var texture = strategy(area, centerTexture);
                        if (texture == centerTexture) continue;

                        centerTexture = texture;
                        handled = true;
                        break;
                    }

                    if (!handled)
                    {
                        foreach (var strategy in strategies)
                        {
                            if (!strategy.CanExecute(area)) continue;

                            centerTexture = strategy.Execute(area);
                            handled = true;
                            break;
                        }
                    }

                    if (!handled) centerTexture = fallbackStrategy.Execute(area);

                    yield return new PointContext(centerOfArea, centerTexture);
                }
        }

        private TextureHolder StrategyForGround(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborThis] == LocationType.Ground) return ground;

            return defaultValue;
        }

        private TextureHolder CoastWithLandToTheNorth(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheNorth;
        }

        private TextureHolder CoastWithLandToTheSouth(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheSouth;
        }

        private TextureHolder CoastWithLandToTheWest(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheWest;
        }

        private TextureHolder CoastWithLandToTheEast(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheEast;
        }

        private TextureHolder CoastWithLandToTheNorthEast(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheNorthEast;
        }

        private TextureHolder CoastWithLandToTheNorthWest(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheNorthWest;
        }

        private TextureHolder CoastWithLandToTheSouthEast(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheSouthEast;
        }

        private TextureHolder CoastWithLandToTheSouthWest(LocationType[] neighbors, TextureHolder defaultValue)
        {
            if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborSouth] == LocationType.Water) return defaultValue;
            if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

            return water.CoastWithLandToTheSouthWest;
        }

        /// <summary>
        /// Holds together geopoint and texture used for that point to draw it.
        /// </summary>
        public struct PointContext
        {
            public readonly GeoPoint GeoPoint;
            public readonly TextureHolder Texture;

            public PointContext(GeoPoint geopoint, TextureHolder texture)
            {
                GeoPoint = geopoint;
                Texture = texture;
            }
        }
    }
}

File: src/v2/core/src/com/mygdx/game/view/Window.java
package com.mygdx.game.view;

import com.badlogic.gdx.graphics.g2d.SpriteBatch;

import com.badlogic.gdx.graphics.Texture;
import com.mygdx.game.Config;
import com.mygdx.game.domain.CityEntity;
import com.mygdx.game.domain.IslandEntity;
import com.mygdx.game.domain.IslandEntityExtensions;
import com.mygdx.game.domain.LandUnitEntity;
import com.mygdx.game.domain.GeoPoint;
import com.mygdx.game.domain.IntendedMapCentre;
import com.mygdx.game.domain.IslandEntityGenerator;
import com.mygdx.game.resources.ITextureConsumer;
import com.mygdx.game.resources.WaterTextures;
import com.mygdx.game.runtime.GameComponent;
import com.mygdx.game.runtime.IBatchDrawer;

import com.mygdx.game.resources.TextureItem;

import io.vavr.collection.List;
import io.vavr.collection.Seq;
import io.vavr.collection.Map;
import io.vavr.collection.HashMap;

/**
 * Contains all entities which need to be visible in Game window and converts them
 * to textures.
 */
public final class Window implements GameComponent,
                                     IBatchDrawer,
                                     ITextureConsumer
{
    private Map<GeoPoint, LocationType> mapPoints = HashMap.<GeoPoint, LocationType>empty();

    // defines single texture generator strategy
    // area: small array 3*3, where central point defines the point which need th have
    // generated texture, and the rest of array items defines neighbers with indexes as constatns above.
    // if strategy can't generate texture, returns defaultTexture.
    private interface Generator {
        TextureHolder apply(LocationType[] area, TextureHolder defaultTexture);
    }

    private WaterTextures water;
    private TextureHolder city;
    private List<Generator> functionStrategies = List.<Generator>empty();
    private ITileStrategy[] strategies = new ITileStrategy[0];
    private ITileStrategy fallbackStrategy;

    public void Initialize(WaterTextures water, TextureHolder city, ITileStrategy fallbackStrategy, ITileStrategy... strategies) {
        this.water = water;
        this.city = city;

        functionStrategies = functionStrategies
            .append(this::CoastWithLandToTheWest)
            .append(this::CoastWithLandToTheEast)
            .append(this::CoastWithLandToTheNorthEast)
            .append(this::CoastWithLandToTheNorthWest)
            .append(this::CoastWithLandToTheSouthEast)
            .append(this::CoastWithLandToTheSouthWest);

        this.strategies = strategies;
        this.fallbackStrategy = fallbackStrategy;
    }

    public void AddIsland(IslandEntity island) {
        for (var item : IslandEntityExtensions.GeneratePoints(island)) {
            var point = new GeoPoint(item.X, item.Y);
            mapPoints = mapPoints.put(point, LocationType.Ground);
        }
    }

    /**
     * Needs to be invoked later then <see cref="AddIsland(IslandEntity)"/>.
     * @param entity
     */
    public void AddCity(CityEntity entity)
    {
        var point = new GeoPoint(entity.x(), entity.y());
        mapPoints.put(point, LocationType.City);
    }

    /**
     * 
     * @param entity
     */
    public void Include(LandUnitEntity entity)
    {
        var point = new GeoPoint(entity.X(), entity.Y());
        mapPoints.put(point, LocationType.LandUnit);
    }

    /**
     * Generates list of textures for given square, when left bottom corner is provided.
     * @return Sequence contains textures, starting from left bottom corner and returning rows first
     */
    public Seq<PointContext> GetWindow(int lbcx, int lbcy, int width, int height) {
        var result = List.<PointContext>empty();
        for (int dy = 0; dy < height; dy++)
            for (int dx = 0; dx < width; dx++)
            {
                var geox = lbcx + dx;
                var geoy = lbcy + dy;

                var area = new LocationType[9];
                var centerOfArea = new GeoPoint(geox, geoy);
                var relativeCenterOfArea = new GeoPoint(dx, dy);

                for (int i = 0; i < 9; i++) area[i] = LocationType.Water;

                area[Directions.NeighborTopLeft] = mapPoints.get(centerOfArea.TopLeft()).getOrElse(LocationType.Water);
                area[Directions.NeighborNorth] = mapPoints.get(centerOfArea.Top()).getOrElse(LocationType.Water);
                area[Directions.NeighborTopRight] = mapPoints.get(centerOfArea.TopRight()).getOrElse(LocationType.Water);
                area[Directions.NeighborWest] = mapPoints.get(centerOfArea.Left()).getOrElse(LocationType.Water);
                area[Directions.NeighborThis] = mapPoints.get(centerOfArea).getOrElse(LocationType.Water);
                area[Directions.NeighborEast] = mapPoints.get(centerOfArea.Right()).getOrElse(LocationType.Water);
                area[Directions.NeighborDownLeft] = mapPoints.get(centerOfArea.DownLeft()).getOrElse(LocationType.Water);
                area[Directions.NeighborSouth] = mapPoints.get(centerOfArea.Down()).getOrElse(LocationType.Water);
                area[Directions.NeighborDownRight] = mapPoints.get(centerOfArea.DownRight()).getOrElse(LocationType.Water);

                var centerTexture = new TextureHolder();
                var handled = false;
                for (var strategy : functionStrategies) {
                    var texture = strategy(area, centerTexture);
                    if (texture == centerTexture) continue;

                    centerTexture = texture;
                    handled = true;
                    break;
                }

                if (!handled) {
                    for (var strategy : strategies) {
                        if (!strategy.CanExecute(area)) continue;

                        centerTexture = strategy.Execute(area);
                        handled = true;
                        break;
                    }
                }

                if (!handled) centerTexture = fallbackStrategy.Execute(area);

                result = result.append(new PointContext(relativeCenterOfArea, centerTexture));
            }
        return result;
    }

    private TextureHolder CoastWithLandToTheWest(LocationType[] neighbors, TextureHolder defaultValue)
    {
        if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

        return water.getCoastWithLandToTheWest();
    }

    private TextureHolder CoastWithLandToTheEast(LocationType[] neighbors, TextureHolder defaultValue)
    {
        if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

        return water.getCoastWithLandToTheEast();
    }

    private TextureHolder CoastWithLandToTheNorthEast(LocationType[] neighbors, TextureHolder defaultValue)
    {
        if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborNorth] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

        return water.CoastWithLandToTheNorthEast;
    }

    private TextureHolder CoastWithLandToTheNorthWest(LocationType[] neighbors, TextureHolder defaultValue)
    {
        if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborNorth] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborSouth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

        return water.CoastWithLandToTheNorthWest;
    }

    private TextureHolder CoastWithLandToTheSouthEast(LocationType[] neighbors, TextureHolder defaultValue)
    {
        if (neighbors[Directions.NeighborWest] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborEast] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborSouth] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

        return water.CoastWithLandToTheSouthEast;
    }

    private TextureHolder CoastWithLandToTheSouthWest(LocationType[] neighbors, TextureHolder defaultValue)
    {
        if (neighbors[Directions.NeighborWest] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborEast] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborNorth] != LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborSouth] == LocationType.Water) return defaultValue;
        if (neighbors[Directions.NeighborThis] != LocationType.Water) return defaultValue;

        return water.CoastWithLandToTheSouthWest;
    }

    public void Initialize()
    {
        // temporar variables to keep sample textures for demo purposes.
        var island = IslandEntityGenerator.Random(new GeoPoint(20, 20));

        this.AddIsland(island);
        this.AddCity(new CityEntity(20, 20));
        this.Include(new LandUnitEntity(21, 20));
    }

    Texture terrainTexture;
    Texture desertRatsTextures;

    public void OnLoaded(Texture texture, TextureItem item) {
        if (item == TextureItem.TERRAIN) terrainTexture = texture;
        if (item == TextureItem.DESERT_RATES) desertRatsTextures = texture;
    }

    public void LoadFinished()
    {
        var waterTextures = new WaterTextures(terrainTexture);
        var cityTexture = new TextureHolder(terrainTexture, new Rectangle(7 * Config.SpriteSize, 9 * Config.SpriteSize, Config.SpriteSize, Config.SpriteSize));
        var groundTexture = new TextureHolder(terrainTexture, new Rectangle(0 * Config.SpriteSize, 0, Config.SpriteSize, Config.SpriteSize));
        var landUnitTexture = new TextureHolder(desertRatsTextures, new Rectangle(1 + 0 * Config.SpriteSize, 1 + 0, Config.SpriteSize, Config.SpriteSize));

        Initialize(waterTextures, cityTexture, new DefaultStrategy(waterTextures.getSea()),
            new CoastWithLandToTheNorthStrategy(waterTextures.getCoastWithLandToTheNorth()),
            new CoastWithLandToTheSouthStrategy(waterTextures.getCoastWithLandToTheSouth()),
            new GroundStrategy(groundTexture),
            new LandUnitStrategy(landUnitTexture),
            new CityStrategy(cityTexture));
    }

    public IntendedMapCentre IntendedMapCentre;

    public void OnDraw(SpriteBatch spriteBatch) {

        var x = IntendedMapCentre.X;
        var y = IntendedMapCentre.Y;

        // display sample island
        var points = this.GetWindow(0 + x, 0 + y, 30, 30);

        for (var it : points) {
            var position = new Vector2(
                it.GeoPoint.X * Config.SpriteSize,
                it.GeoPoint.Y * Config.SpriteSize);
            spriteBatch.Draw(position, it.Texture);
        }
    }



    /**
     * Holds together geopoint and texture used for that point to draw it.
     */
    public class PointContext {
        public final GeoPoint GeoPoint;
        public final TextureHolder Texture;

        public PointContext(GeoPoint geopoint, TextureHolder texture)
        {
            GeoPoint = geopoint;
            Texture = texture;
        }
    }
}
